File: services/auth-core/src/request_proof.rs
//! DPoP-style per-request proof-of-possession (ADR-0011, R-308).
//!
//! A sender-constrained access token: each authenticated request carries a proof that the caller
//! holds the device's enrolled private key. The proof is a raw ECDSA-P256 signature over this
//! canonical, domain-separated transcript — the same vetted signing path as auth/refresh, so a
//! stolen bearer token is useless without the non-exportable key. RFC 9449 semantics (bind
//! method + URI + token + time + nonce), expressed in Sentinel's transcript idiom rather than JWS.

use crate::crypto::verify_p256;

/// Domain-separation tag. Versioned; a new proof format re-tags.
pub const DOMAIN: &[u8] = b"app.sentinel.dpop.v1";

/// Protocol version carried in the signed bytes (explicit, non-silent evolution).
pub const PROTOCOL_VERSION: u16 = 1;

/// Maximum accepted clock skew (seconds) between the proof timestamp and server time. Bounds the
/// replay window; the nonce cache makes each proof single-use within it.
pub const MAX_SKEW_SECS: u64 = 60;

/// The fields bound into one request proof.
pub struct RequestProof<'a> {
    /// HTTP method, uppercase ASCII (e.g. `GET`, `POST`).
    pub method: &'a [u8],
    /// Request path, no query string (e.g. `/v1/inbox`).
    pub path: &'a [u8],
    /// SHA-256 of the presented access token — recomputed server-side, so the client cannot lie
    /// about which token the proof covers.
    pub access_token_hash: &'a [u8; 32],
    /// Client clock, unix seconds.
    pub timestamp: u64,
    /// 16 random bytes, unique per request (single-use within the skew window).
    pub nonce: &'a [u8; 16],
}

impl<'a> RequestProof<'a> {
    /// The canonical byte string that is signed and verified.
    pub fn encode(&self) -> Vec<u8> {
        let mut out = Vec::with_capacity(
            4 + DOMAIN.len()
                + 2
                + (4 + self.method.len())
                + (4 + self.path.len())
                + (4 + 32)
                + 8
                + (4 + 16),
        );
        put_lp(&mut out, DOMAIN);
        out.extend_from_slice(&PROTOCOL_VERSION.to_be_bytes());
        put_lp(&mut out, self.method);
        put_lp(&mut out, self.path);
        put_lp(&mut out, self.access_token_hash);
        out.extend_from_slice(&self.timestamp.to_be_bytes());
        put_lp(&mut out, self.nonce);
        out
    }

    /// Verify `signature` over this proof against the device's SEC1 public key. Fail-closed.
    pub fn verify(&self, device_public_key_sec1: &[u8], signature: &[u8]) -> bool {
        verify_p256(device_public_key_sec1, &self.encode(), signature)
    }

    /// True iff `timestamp` is within `MAX_SKEW_SECS` of `now` (both directions).
    pub fn is_fresh(&self, now: u64) -> bool {
        now.abs_diff(self.timestamp) <= MAX_SKEW_SECS
    }
}

fn put_lp(out: &mut Vec<u8>, field: &[u8]) {
    out.extend_from_slice(&(field.len() as u32).to_be_bytes());
    out.extend_from_slice(field);
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample(ts: u64) -> Vec<u8> {
        RequestProof {
            method: b"POST",
            path: b"/v1/conversations/aabb/messages",
            access_token_hash: &[7u8; 32],
            timestamp: ts,
            nonce: &[9u8; 16],
        }
        .encode()
    }

    #[test]
    fn golden_vector_is_stable() {
        let hex: String = sample(1_700_000_000)
            .iter()
            .map(|b| format!("{b:02x}"))
            .collect();
        assert_eq!(
            hex,
            "000000146170702e73656e74696e656c2e64706f702e7631000100000004504f53540000001f2f76312f636f6e766572736174696f6e732f616162622f6d65737361676573000000200707070707070707070707070707070707070707070707070707070707070707000000006553f1000000001009090909090909090909090909090909"
        );
    }

    #[test]
    fn encoding_is_injective_across_method_and_path() {
        let get = RequestProof {
            method: b"GET",
            path: b"/v1/inbox",
            access_token_hash: &[0u8; 32],
            timestamp: 1,
            nonce: &[0u8; 16],
        }
        .encode();
        let post = RequestProof {
            method: b"POST",
            path: b"/v1/inbox",
            access_token_hash: &[0u8; 32],
            timestamp: 1,
            nonce: &[0u8; 16],
        }
        .encode();
        assert_ne!(get, post, "method must be bound");
        // Moving a byte from method into path must not collide (length prefixes).
        let a = RequestProof {
            method: b"GE",
            path: b"T/v1/inbox",
            access_token_hash: &[0u8; 32],
            timestamp: 1,
            nonce: &[0u8; 16],
        }
        .encode();
        assert_ne!(get, a);
    }

    #[test]
    fn freshness_window() {
        let p = RequestProof {
            method: b"GET",
            path: b"/v1/inbox",
            access_token_hash: &[0u8; 32],
            timestamp: 1000,
            nonce: &[0u8; 16],
        };
        assert!(p.is_fresh(1000));
        assert!(p.is_fresh(1000 + MAX_SKEW_SECS));
        assert!(p.is_fresh(1000 - MAX_SKEW_SECS));
        assert!(!p.is_fresh(1000 + MAX_SKEW_SECS + 1));
        assert!(!p.is_fresh(1000 - MAX_SKEW_SECS - 1));
    }

    #[test]
    fn sign_verify_round_trip_and_tamper() {
        use p256::ecdsa::{signature::Signer, Signature, SigningKey};
        let signing = SigningKey::random(&mut rand_core::OsRng);
        let public = signing
            .verifying_key()
            .to_encoded_point(false)
            .as_bytes()
            .to_vec();
        let proof = RequestProof {
            method: b"POST",
            path: b"/v1/inbox/ack",
            access_token_hash: &[3u8; 32],
            timestamp: 42,
            nonce: &[5u8; 16],
        };
        let sig: Signature = signing.sign(&proof.encode());
        assert!(proof.verify(&public, &sig.to_bytes()));

        // A different path (same signature) must not verify — request binding holds.
        let other = RequestProof {
            path: b"/v1/inbox",
            ..RequestProof {
                method: b"POST",
                path: b"/v1/inbox/ack",
                access_token_hash: &[3u8; 32],
                timestamp: 42,
                nonce: &[5u8; 16],
            }
        };
        assert!(!other.verify(&public, &sig.to_bytes()));
    }
}

File: apps/ios/NedwonsKit/Sources/NedwonsKit/RequestProof.swift
import CryptoKit
import Foundation

/// DPoP-style per-request proof-of-possession (ADR-0011, R-308). The Swift encoder reproduces
/// `auth_core::request_proof::RequestProof::encode` **byte-for-byte** so the device signs exactly
/// what the server verifies. A sender-constrained access token: a stolen bearer token is useless
/// without the enrolled device key that signs each request.
public struct RequestProof: Sendable {
    public let method: String
    /// Request path, no query string (e.g. `/v1/inbox`).
    public let path: String
    /// SHA-256 of the presented access token (bytes).
    public let accessTokenHash: Data
    public let timestamp: UInt64
    public let nonce: Data  // 16 bytes

    public init(method: String, path: String, accessTokenHash: Data, timestamp: UInt64, nonce: Data)
    {
        self.method = method
        self.path = path
        self.accessTokenHash = accessTokenHash
        self.timestamp = timestamp
        self.nonce = nonce
    }

    static let domain = Data("app.nedwons.dpop.v1".utf8)
    static let protocolVersion: UInt16 = 1

    /// The canonical byte string that is signed and verified (matches the Rust encoder).
    public func canonicalBytes() -> Data {
        var out = Data()
        Self.putLengthPrefixed(&out, Self.domain)
        Self.putUInt16(&out, Self.protocolVersion)
        Self.putLengthPrefixed(&out, Data(method.utf8))
        Self.putLengthPrefixed(&out, Data(path.utf8))
        Self.putLengthPrefixed(&out, accessTokenHash)
        Self.putUInt64(&out, timestamp)
        Self.putLengthPrefixed(&out, nonce)
        return out
    }

    /// Build the `X-Nedwons-Proof` header value for a request: hashes the access token, signs the
    /// canonical proof with `signer` (the private key never leaves the signer), and formats the
    /// `v1;ts=…;nonce=…;sig=…` header. `nonce` defaults to 16 fresh random bytes (single-use).
    public static func header(
        signer: DeviceSigner,
        accessToken: Data,
        method: String,
        path: String,
        timestamp: UInt64 = UInt64(Date().timeIntervalSince1970),
        nonce: Data = Data((0..<16).map { _ in UInt8.random(in: 0...255) })
    ) throws -> String {
        let tokenHash = Data(SHA256.hash(data: accessToken))
        let proof = RequestProof(
            method: method, path: path, accessTokenHash: tokenHash, timestamp: timestamp,
            nonce: nonce)
        let signature = try signer.sign(proof.canonicalBytes())
        return "v1;ts=\(timestamp);nonce=\(Hex.encode(nonce));sig=\(Hex.encode(signature))"
    }

    private static func putLengthPrefixed(_ out: inout Data, _ field: Data) {
        putUInt32(&out, UInt32(field.count))
        out.append(field)
    }
    private static func putUInt16(_ out: inout Data, _ value: UInt16) {
        withUnsafeBytes(of: value.bigEndian) { out.append(contentsOf: $0) }
    }
    private static func putUInt32(_ out: inout Data, _ value: UInt32) {
        withUnsafeBytes(of: value.bigEndian) { out.append(contentsOf: $0) }
    }
    private static func putUInt64(_ out: inout Data, _ value: UInt64) {
        withUnsafeBytes(of: value.bigEndian) { out.append(contentsOf: $0) }
    }

    /// The exact input behind the Rust golden vector, for cross-platform byte-identity testing.
    public static func sampleVectorInput() -> RequestProof {
        RequestProof(
            method: "POST",
            path: "/v1/conversations/aabb/messages",
            accessTokenHash: Data(repeating: 0x07, count: 32),
            timestamp: 1_700_000_000,
            nonce: Data(repeating: 0x09, count: 16)
        )
    }
}
